layerfetcher/layers.go

```diff
@@ -237,208 +250,232 @@ func (l *Logic) tortoiseBeaconReceiver(ctx context.Context, msg []byte) []byte {
-	for hash, peers := range hashes {
+	if numErrors > numPeers/2 {
+		logger.With().Error("too many errors from peers",
+			layerID,
+			log.Int("numPeers", numPeers),
+			log.Int("numErrors", numErrors))
+		result.Hashes = nil
+		result.Err = ErrTooManyPeerErrors
+	}
+	for _, ch := range channels {
+		ch <- result
+	}
+}
+
+// PollLayerBlocks polls peers on blocks for given layer hashes. for each layer hash, it requests the blocks from a
+// peer that reported the layer hash previously. it returns a channel for the caller to be notified when responses
+// are received from all peers.
+func (l *Logic) PollLayerBlocks(ctx context.Context, layerID types.LayerID, hashes map[types.Hash32][]peers.Peer) chan LayerPromiseResult {
+	resChannel := make(chan LayerPromiseResult, 1)
+	l.mutex.Lock()
+	l.layerBlocksChs[layerID] = append(l.layerBlocksChs[layerID], resChannel)
+	if _, ok := l.layerBlocksRes[layerID]; ok {
+		// polling of block hashes of  is on-going, just wait for the polling to finish and get notified
+		l.mutex.Unlock()
+		return resChannel
+	}
+	l.layerBlocksRes[layerID] = make(map[peers.Peer]*peerResult)
+	l.mutex.Unlock()
+
+	// send a request to the first peer of the list to get blocks data.
+	// todo: think if we should aggregate or ask from multiple peers to have some redundancy in requests
+	numHashes := len(hashes)
+	for hash, remotePeers := range hashes {
+		peer := remotePeers[0]
 		if hash == emptyHash {
-			l.receiveBlockHashes(ctx, id, nil, len(hashes), ErrZeroLayer)
+			l.receiveBlockHashes(ctx, layerID, peer, numHashes, nil, ErrZeroLayer)
 			continue
 		}
-		// build receiver function
 		receiveForPeerFunc := func(data []byte) {
-			l.receiveBlockHashes(ctx, id, data, len(hashes), nil)
+			l.receiveBlockHashes(ctx, layerID, peer, numHashes, data, nil)
 		}
 		errFunc := func(err error) {
-			l.receiveBlockHashes(ctx, id, nil, len(hashes), err)
+			l.receiveBlockHashes(ctx, layerID, peer, numHashes, nil, err)
 		}
-		err := l.net.SendRequest(ctx, server.LayerBlocksMsg, hash.Bytes(), peers[0], receiveForPeerFunc, errFunc)
+		err := l.net.SendRequest(ctx, server.LayerBlocksMsg, hash.Bytes(), peer, receiveForPeerFunc, errFunc)
 		if err != nil {
-			l.receiveBlockHashes(ctx, id, nil, len(hashes), err)
+			l.receiveBlockHashes(ctx, layerID, peer, numHashes, nil, err)
 		}
 	}
-
+	return resChannel
 }
 
-func (l *Logic) determinePromiseResult(layer types.LayerID, errs []error) *LayerPromiseResult {
-	hasZeroBlockHash := false
-	for _, e := range errs {
-		if e == nil {
-			// at least one layer hash contains blocks. not a zero block layer
-			return &LayerPromiseResult{Layer: layer, Err: nil}
+// parseBlockHashAndFetch parses the block IDs from a peer's response and fetch the corresponding blocks.
+// it returns a peerResult based on the response.
+func (l *Logic) parseBlockHashAndFetch(ctx context.Context, layerID types.LayerID, data []byte, err error) *peerResult {
```

Comment:
rename to `parseBlocksHashAndFetch`

Reply:
personally, and as part of SOLID design I think this method should be split into 2, the name suggest it does 2 different operations -parse and fetch, IMO parsing can be done in calling code

Reply:
good idea. i moved the parsing to the calling code. thanks